pkg/server/sandbox_portforward_unix.go

```diff
@@ -62,48 +61,64 @@ func (c *criService) portForward(ctx context.Context, id string, port int32, str
 		netNSPath = "host"
 	}
 
-	socat, err := exec.LookPath("socat")
-	if err != nil {
-		return errors.Wrap(err, "failed to find socat")
-	}
-
-	// Check https://linux.die.net/man/1/socat for meaning of the options.
-	args := []string{socat, "-", fmt.Sprintf("TCP4:localhost:%d", port)}
-
-	log.G(ctx).Infof("Executing port forwarding command %q in network namespace %q", strings.Join(args, " "), netNSPath)
+	log.G(ctx).Infof("Executing port forwarding in network namespace %q", netNSPath)
 	err = netNSDo(func(_ ns.NetNS) error {
-		cmd := exec.Command(args[0], args[1:]...)
-		cmd.Stdout = stream
-
-		stderr := new(bytes.Buffer)
-		cmd.Stderr = stderr
-
-		// If we use Stdin, command.Run() won't return until the goroutine that's copying
-		// from stream finishes. Unfortunately, if you have a client like telnet connected
-		// via port forwarding, as long as the user's telnet client is connected to the user's
-		// local listener that port forwarding sets up, the telnet session never exits. This
-		// means that even if socat has finished running, command.Run() won't ever return
-		// (because the client still has the connection and stream open).
-		//
-		// The work around is to use StdinPipe(), as Wait() (called by Run()) closes the pipe
-		// when the command (socat) exits.
-		in, err := cmd.StdinPipe()
+		defer stream.Close()
+		// TODO: hardcoded to tcp4 because localhost resolves to ::1 by default if the system has IPv6 enabled.
+		// Theoretically happy eyeballs will try IPv6 first and fallback to IPv4
+		// but resolving localhost doesn't seem to return and IPv4 address, thus failing the connection.
+		conn, err := net.Dial("tcp4", fmt.Sprintf("localhost:%d", port))
```

Comment:
@aojea could we do something like this?
```suggestion
		conn, err := net.Dial("tcp4", fmt.Sprintf("%s:%d", s.IP, port))
```

Reply:
I was just keeping the old behaviour because of my ignorance on how the ports are exposed inside the namespace.

I can explain at the network level the problem and why I think is forcing IPv4 here.
`localhost`, depending on some factors, may resolve to `[::1]` or `127.0.0.1`, Long history short, commonly it resolves to [::1] if you are more interested in the whole history I explain it in this link https://gist.github.com/aojea/94f6f483173641647c731f582e52f0b0

But typically, the applications use to listen only 127.0.0.1 inside the container :shrug: , that's the part I'm missing. 

If we are able to identify correctly the address that the app is listening, then is better using ti, so this works for both IPv4 and IPv6, but I prefer having @mikebrow guidance here.

Reply:
it's hardcoded localhost(or 127.0.0.1):random port for sec. reasons... or at least that's what the decision was way back when.

Any change here would likely need a kep that sets up streaming ip(s) and port(s) at the CRI request and/or response level.